src/index.js

```diff
@@ -1,37 +1,46 @@
 'use strict'
 
-const EventEmitter = require('events')
-const pull = require('pull-stream/pull')
-const empty = require('pull-stream/sources/empty')
-const asyncEach = require('async/each')
-const TimeCache = require('time-cache')
+const assert = require('assert')
 const debug = require('debug')
+const EventEmitter = require('events')
 const errcode = require('err-code')
 
-const Peer = require('./peer')
+const PeerInfo = require('peer-info')
+
 const message = require('./message')
+const Peer = require('./peer')
+const utils = require('./utils')
 const {
   signMessage,
   verifySignature
 } = require('./message/sign')
-const utils = require('./utils')
-
-const nextTick = require('async/nextTick')
 
 /**
  * PubsubBaseProtocol handles the peers and connections logic for pubsub routers
  */
 class PubsubBaseProtocol extends EventEmitter {
   /**
-   * @param {String} debugName
-   * @param {String} multicodec
-   * @param {Object} libp2p libp2p implementation
-   * @param {Object} options
-   * @param {boolean} options.signMessages if messages should be signed, defaults to true
-   * @param {boolean} options.strictSigning if message signing should be required, defaults to true
-   * @constructor
+   * @param {String} debugName log namespace
+   * @param {Array<string>|string} multicodecs protocol identificers to connect
+   * @param {PeerInfo} peerInfo peer's peerInfo
+   * @param {Object} registrar registrar for libp2p protocols
+   * @param {function} registrar.register
+   * @param {function} registrar.unregister
+   * @param {Object} [options]
+   * @param {boolean} [options.signMessages] if messages should be signed, defaults to true
+   * @param {boolean} [options.strictSigning] if message signing should be required, defaults to true
+   * @abstract
    */
-  constructor (debugName, multicodec, libp2p, options) {
+  constructor (debugName, multicodecs, peerInfo, registrar, options = {}) {
```

Comment:
You should be considering the Pubsub implementation instead

Take a look at: https://github.com/libp2p/js-libp2p-pubsub/pull/26/files#diff-04c6e90faac2675aa89e2176d2eec7d8R54

Since this class is not supposed to be used directly, I did not document the constructor in the README

Reply:
Should I add to the `Readme.md` that this is an abstract class?

Reply:
I think it's probably ok, I was just looking at the implementation constructor instead of the call.